### PR TITLE
Fix the E8M0 scale computation in the MXFP4 (W4A4) MOE CUTLASS kernel

### PR DESCRIPTION
**Pull request by xin3he**
## Purpose

Fix the E8M0 scale computation in the MXFP4 (W4A4) MOE CUTLASS kernel, which
currently produces severely incorrect quantization results.

Related PR: #37463

## Root Cause

The original code computes the E8M0 scale factor via:

```c++
// Old (buggy):
float SFValue = vecMax * reciprocal_approximate_ftz(6.0f);  // vecMax / 6.0
uint32_t tmp = reinterpret_cast<uint32_t&>(SFValue) >> 23;
fp8SFVal = tmp & 0xff;  // scale_exp = biased_exponent(vecMax / 6.0)
```

This is **mathematically wrong** because:

floor(log2(a / b)) != floor(log2(a)) - floor(log2(b))

Dividing by 6.0 *before* extracting the exponent shifts the mantissa boundary,
causing the resulting exponent to be 1 too small for a large fraction of inputs.
When the scale is too small, `vecMax / scale` far exceeds 6.0 (the E2M1 max),
and the block-max value saturates -- destroying information.

| vecMax | Old: `biased_exp(vecMax/6)` | Correct: `biased_exp(vecMax)-2` | Ratio (old) | Ratio (fixed) |
|--------|-------|-------|------|------|
| 2.7 | 125 (scale=0.25) | 126 (scale=0.5) | 10.8  | 5.4 |
| 5.5 | 126 (scale=0.5) | 127 (scale=1.0) | 11.0  | 5.5 |
| 1.2 | 124 (scale=0.125) | 125 (scale=0.25) | 9.6  | 4.8  |

## Fix: Exponent-First with 0.75-Threshold Rounding

The new implementation extracts the exponent directly from `vecMax`, applies a
rounding bias, then subtracts 2:

```c++
uint32_t max_bits = __float_as_uint(vecMax);
// Round up exponent when mantissa >= 0.75
uint32_t rounded_bits = (max_bits + (1u << 21)) & 0xFF800000u;
uint32_t biased_exp = (rounded_bits >> 23) & 0xFFu;
uint32_t scale_exp = (biased_exp > 2u) ? (biased_exp - 2u) : 0u;
```

### Origin: Matching the Quantization Toolchain

This rounding scheme (`VAL_TO_ADD = 1 << (mantissa_bits - 2)`) is used by upstream quantization libraries:

- **compressed-tensors** ([`quantization/utils/mxfp_utils.py`](https://github.com/vllm-project/compressed-tensors/blob/c170de622fbbed42f6792c1ca4397cc76a960225/src/compressed_tensors/quantization/utils/mxfp_utils.py#L113-L118)):
  ```python
  VAL_TO_ADD = 1 << (mantissa - FP4_E2M1_DATA.mantissa - 1)  # 1<<21 for fp32
  block_max_uint = torch.bitwise_and(x + VAL_TO_ADD, SIGN_EXPONENT_MASK)
  scale_exp = 127 + torch.floor(torch.log2(block_max)) - 2
  ```

The CUDA kernel **must** match these implementations to ensure consistency
between offline weight quantization and online activation quantization.

### Why This Helps Precision (Not Strict OCP MX)

This is **not** the strict OCP MX spec. The differences:

| Approach | Formula | Clipping zone | Behavior |
|----------|---------|---------------|----------|
| **Strict OCP MX** | `biased_exp(vecMax) - 2` | mantissa in  [0.5, 1.0) -> ratio in [6, 8) | 50% of max values clip |
| **This fix (threshold 0.75)** | `biased_exp(round_0.75(vecMax)) - 2` | mantissa in [0.5, 0.75) -> ratio in [6, 7) | 25% of max values clip |
| **Threshold 0.5 (1<<22)** | `biased_exp(round_0.5(vecMax)) - 2` | never clips | scale often too large |

The 0.75 threshold is a **precision-optimized tradeoff**:

1. **Avoids premature scale increase**: For mantissa in [0.5, 0.75), the code
   keeps the smaller scale rather than doubling it. If the scale doubles, ALL
   values in the 16-element block lose half their effective resolution.

2. **Tolerates mild clipping of block-max**: The ratio falls in [6.0, 7.0),
   so vecMax clips to 6.0 -- at most ~14% error on a single value. This is a
   better tradeoff than halving the precision of all 16 values.

3. **Prevents extreme clipping**: For mantissa >= 0.75 (ratio would be >= 7.0),
   the exponent rounds up, preventing excessive saturation.

In summary: the dynamic range is effectively extended to [0, 7*scale) with
graceful saturation at 6.0, **avoiding the scale growing too early** which
would compress the representable range for the entire block.

## Other Fixes

- **Input contiguity**: Added `.contiguous()` for FlashInfer quantization path
  to prevent incorrect results from non-contiguous tensors.

## Test Plan

- UT:
```bash
CUDA_VISIBLE_DEVICES=7 .venv/bin/python -m pytest tests/kernels/moe/test_mxfp4_moe.py -v
```

- Real model:
Verified on Kimi MXFP4 model
https://huggingface.co/INCModel/Kimi-K2.5-MXFP4-AutoRound

## Test Result

Command (B200): 
```bash
CUDA_VISIBLE_DEVICES=4,5,6,7 lm_eval --model vllm     --model_args pretrained=/workspace/models/moonshotai/Kimi-K2.6,tensor_parallel_size=4,data_parallel_size=1,max_model_len=8192,trust_remote_code=True --tasks gsm8k,mmlu,piqa,hellaswag  --batch_size auto
```

lm_eval on B200 (Kimi-K2.6, TP=4):

| Config | gsm8k | mmlu | piqa | hellaswag | avg |
|--------|-------|------|------|-----------|-----|
| FP16 weights (W4A16) | 0.948 | 0.889 | 0.852 | 0.773 | 0.866 |
| MXFP4 W4A4 (old, broken) | 0.809 | 0.808 | 0.809 | 0.653 | 0.770 |
| MXFP4 W4A4 (**this fix**) | 0.921 | 0.874 | 0.842 | 0.744 | 0.845 |

Recovers ~78% of the quantization accuracy loss.

cc @dsikka @mgoin @yiliu30 @wenhuach21 @thuang6 @hshen14

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by xin3he** [comment]
Working on the code review from gemini.

File: csrc/libtorch_stable/quantization/fp4/nvfp4_utils.cuh [hunk 1]
```diff
@@ -237,21 +237,30 @@ __device__ __forceinline__ fp4_packed_t cvt_warp_fp16_to_fp4(
...
+    fp8SFVal = static_cast<uint8_t>(scale_exp);
+    // Reconstruct scale as float32: scale = 2^(scale_exp - 127)
+    uint32_t sf_bits = scale_exp << 23;
+    SFValue = __uint_as_float(sf_bits);
   } else {
+    // NVFP4 path: scale = max / 6.0, stored as E4M3.
+    SFValue = SFScaleVal * (vecMax * reciprocal_approximate_ftz(6.0f));
     // Here SFValue is always positive, so E4M3 is the same as UE4M3.
     __nv_fp8_e4m3 tmp = __nv_fp8_e4m3(SFValue);
     reinterpret_cast<__nv_fp8_e4m3&>(fp8SFVal) = tmp;
```
[the remark this answers is not in the document]
Comment (xin3he) [reply csrc/libtorch_stable/quantization/fp4/nvfp4_utils.cuh:266]: Have `SFValue = float(tmp);`

File: csrc/libtorch_stable/quantization/fp4/nvfp4_utils.cuh [hunk 2]
```diff
@@ -237,21 +237,30 @@ __device__ __forceinline__ fp4_packed_t cvt_warp_fp16_to_fp4(
...
-    reinterpret_cast<uint32_t&>(SFValue) = tmp << 23;
+    // OCP MX spec E8M0 scale computation (MXFP4 path):
+    // scale_exp = biased_exponent(round_up(vecMax)) - 2
+    //   -2 because max E2M1 value is 6.0 ≈ 2^2.58; we use 2^2=4 as the
+    //   safe divisor so that max_val / scale <= 6.0 for values near 2^n.
+    uint32_t max_bits = __float_as_uint(vecMax);
+    // Add rounding bias at mantissa bit 21 (equivalent to bf16 val_to_add=32
+    // at bit 5). Threshold: values with mantissa >= 0.75 (i.e. >= 1.75*2^n)
+    // round up to the next power of 2.
+    uint32_t rounded_bits = (max_bits + (1u << 21)) & 0xFF800000u;
```
[the remark this answers is not in the document]
Comment (xin3he) [reply csrc/libtorch_stable/quantization/fp4/nvfp4_utils.cuh:253]: Refused. This will cause a huge drop of accuracy

**Comment by hshen14** [comment]
@dsikka @mgoin Please help review. Thanks.

**Comment by jikunshang** [comment]
can you run this case provided by  https://github.com/vllm-project/vllm/pull/37463
```
vllm serve nm-testing/Qwen3-30B-A3B-MXFP4A16
python tests/evals/gsm8k/gsm8k_eval.py
```

**Comment by xin3he** [comment]
Thanks @jikunshang , Here is the additional test:
```bash
vllm serve nm-testing/Qwen3-30B-A3B-MXFP4A16
python tests/evals/gsm8k/gsm8k_eval.py
```
```log
Results:
Accuracy: 0.857
Invalid responses: 0.000
Total latency: 21.832 s
Questions per second: 60.415
Total output tokens: 154938
Output tokens per second: 7096.714
```
vs results of the original w4a4 kernel from https://github.com/vllm-project/vllm/pull/37463
```log
Results:
Accuracy: 0.835
Invalid responses: 0.000
Total latency: 24.276 s
Questions per second: 54.333
Total output tokens: 150954
Output tokens per second: 6218.226
```

**Comment by jikunshang** [comment]
thanks, please fix pre-commit issue.

**Comment by xin3he** [comment]
@brian-dellabetta Here I have submitted a PR aimed at resolving the MXFP4 precision issues related to LLMC. I would be very grateful if you could assist in moving this PR forward.

**Review by jikunshang — APPROVED** [review]
LGTM since accuracy got improved. 
@mgoin  @yewentao256  @zyongye  please take a look, thanks!

**Comment by xin3he** [comment]
Seems like those failures are not related to this PR.

**Comment by xin3he** [comment]
@jikunshang 
I'd like to get your help about the CI failure. There are two failures:
- [amd-entrypoints-integration-api-server-openai](https://buildkite.com/vllm/ci/builds/70697#019eaa72-2614-4d4c-af89-9dca1e83b63f) it passes in my local env.
- [kernels-fusedmoe-layer-test-2-b200s](https://buildkite.com/vllm/ci/builds/70697#019eaa72-263a-4fe5-95ba-b80cdda9d8a3) It seems like a timeout issue.

I had merged main to trigger the CI many times to avoid the random issue, but the random issues continue happens. 
The failures are different in each re-run.

**Review by mgoin — APPROVED** [review]

**Comment by mgoin** [comment]
Great work!
